src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -5716,12 +5716,23 @@ private bool IsOpenName()
                             explicitInterfaceName,
                             AddError(separator, ErrorCode.ERR_ExplicitEventFieldImpl));
 
-                        Debug.Assert(beforeIdentifierPointSet);
-                        Reset(ref beforeIdentifierPoint);
-
-                        //clear fields that were populated after the reset point
-                        identifierOrThisOpt = null;
-                        typeParameterListOpt = null;
+                        /*
+                         * It feels like this recovery is applied too aggressively. For example, if there is no line break between the last
+                         * dot and the last identifier, it feels reasonable to assume that both belong to the same declaration (the event 
+                         * declaration), rather than the identifier belongs to a declaration "below" an incomplete explicit interface 
+                         * implementation. It might be worth trying to restrict the recovery that drops the name only to cases when there 
+                         * is a line break between the tokens. Also, if the identifier is followed by a ;, we probably shouldn't apply the
+                         * recovery even when there is a line break between the tokens.
+                         */
+
+                        if (separator.HasTrailingTrivia && this.CurrentToken.Kind != SyntaxKind.SemicolonToken)
```

Comment:
>separator.HasTrailingTrivia [](start = 28, length = 27)

The presence of trailing trivia doesn't really mean there is a line break, it could be just a space (please make sure you have a test for a scenario like this). I think we should check for ```TrailingTrivia.Any((int)SyntaxKind.EndOfLineTrivia)```, see other places in this file where we check for a line break.